pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -864,18 +866,13 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
 				return volumeToMount.GenerateError("MapVolume failed", fmt.Errorf("Device path of the volume is empty"))
 			}
 		}
-		// Set up global map path under the given plugin directory using symbolic link
-		globalMapPath, err :=
-			blockVolumeMapper.GetGlobalMapPath(volumeToMount.VolumeSpec)
-		if err != nil {
-			// On failure, return error. Caller will log and retry.
-			return volumeToMount.GenerateError("MapVolume.GetDeviceMountPath failed", err)
-		}
+
 		mapErr = og.blkUtil.MapDevice(devicePath, globalMapPath, string(volumeToMount.Pod.UID))
```

Comment:
`len(globalMapPath)` should be checked here. If globalMapPath is empty, we have to call blockVolumeMapper.GetGlobalMapPath() for non-attachable plugin too. Otherwise MapDevice fails.

Reply:
I see. One question is what is the purpose of blockVolumeMapper.SetUpDevice()? All block volume plugins do nothing and returns empty string.

Reply:
If plugin is non-attachable, the plugin should implement SetUpDevice() to attach and mount volume to kubelet node. This is the same purpose of SetUp() for Filesystem volume.

>> All block volume plugins do nothing and returns empty string.

yeah, currently we only have attachable plugins for block, so the method is empty.

Reply:
A little bit confusing for me. If SetUpDevice() is also attach volume to kubelet node, what is the difference between attachable and non-attachable? For attachable plugins since SetUpDevice does nothing, we don't need to call it at all. Also the current workflow is that for attachable plugins, 
1. waitForAttach(): attach volume
2. SetUpDevice(): does nothing
3. MapDevice(): called twice
4. AttachFileDevice()
Why not put at least 3 and 4 into same volume plugin function? If possible, I think we should just put one function for volume.Setup() and one function for volume.Teardown() since all the steps are done one the node side.

Reply:
@jingxu97 
Since 3 and 4 are common operations for both attachable and non-attachable plugins, we decided to implement these in operation generator instead of in each plugin side.

- 3  MapDevice(): create symbolic link for global map path and pod device map path
- 4  AttachFileDevice(): Create loop device for attached volume on kubelet node

As a result, attachable plugins don't need anything in SetUpDevice(), and non-attachable plugin only do attach during SetUpDevice().
Other operations are taken care by kubelet side.

@msau42
Any additional comments?